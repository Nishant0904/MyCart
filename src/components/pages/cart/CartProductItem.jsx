// import { useState } from 'react'
import styled from 'styled-components'
import { Link } from 'react-router-dom'
import formatToUSD from '../../../utils/formatToUSD'
import PropTypes from 'prop-types'
import ButtonIcon from '../../reusables/buttons/ButtonIcon'
import trashIcon from '../../../assets/icons/trash-2.svg'

const CartProductItem = ({ product, handleDeleteFromCart }) => {
    return (
        <StyledCartProductItem data-testid='cart-product-item'>
            <Link
                to={`/products/${product.id}`}
                data-testid='cart-product-item__image-link'
            >
                <img src={product.image} alt={product.title} />
            </Link>
            <ProductItemInfo>
                <div>
                    <Link
                        to={`/products/${product.id}`}
                        data-testid='cart-product-item__title-link'
                    >
                        <h4>{product.title}</h4>
                    </Link>
                    <p>{formatToUSD(product.price)}</p>
                </div>
                <ProductItemButtonsContainer>
                    <p data-testid='cart-product-item__quantity'>
                        Qty: {product.quantity}
                    </p>
                    <ButtonIcon>
                        <img
                            src={trashIcon}
                            alt='trash icon'
                            onClick={() => {
                                handleDeleteFromCart(product)
                            }}
                            title='Delete all of this item'
                            data-testid='cart-product-item__trash-btn'
                        />
                    </ButtonIcon>
                </ProductItemButtonsContainer>
            </ProductItemInfo>
        </StyledCartProductItem>
    )
}

const StyledCartProductItem = styled.li`
    display: flex;
    gap: 16px;
    padding: 12px 16px;

    a {
        cursor: pointer;
    }

    img {
        max-width: 96px;
    }

    button {
        align-self: start;
    }
`

const ProductItemInfo = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    gap: 24px;
`

const ProductItemButtonsContainer = styled.div`
    display: flex;
    align-items: center;
    gap: 8px;
`

CartProductItem.propTypes = {
    product: PropTypes.object.isRequired,
    handleDeleteFromCart: PropTypes.func.isRequired,
}

export default CartProductItem
