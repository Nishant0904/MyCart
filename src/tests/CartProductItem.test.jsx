import { describe, it, expect, vi } from 'vitest'
import { render, screen, waitFor } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import CartProductItem from '../components/pages/cart/CartProductItem'
import userEvent from '@testing-library/user-event'

const mockProduct = {
    id: 1,
    title: 'Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops',
    price: 109.95,
    description:
        'Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday',
    category: "men's clothing",
    image: 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg',
    rating: {
        rate: 3.9,
        count: 120,
    },
    quantity: 1,
}

describe('Cart Product Item test', () => {
    it('renders correctly', () => {
        render(
            <CartProductItem
                product={mockProduct}
                handleDeleteFromCart={() => {}}
            />,
            { wrapper: BrowserRouter },
        )
        expect(screen.getByTestId('cart-product-item')).toBeInTheDocument()
        expect(screen.getByTestId('cart-product-item')).toMatchSnapshot()
    })

    it('image link click takes user to product page', async () => {
        const user = userEvent.setup()
        render(
            <CartProductItem
                product={mockProduct}
                handleDeleteFromCart={() => {}}
            />,
            { wrapper: BrowserRouter },
        )
        const imageLink = screen.getByTestId('cart-product-item__image-link')

        await user.click(imageLink)
        await waitFor(() => {
            expect(window.location.pathname).toBe(`/products/${mockProduct.id}`)
        })
    })

    it('title link click takes user to product page', async () => {
        const user = userEvent.setup()
        render(
            <CartProductItem
                product={mockProduct}
                handleDeleteFromCart={() => {}}
            />,
            { wrapper: BrowserRouter },
        )
        const titleLink = screen.getByTestId('cart-product-item__title-link')

        await user.click(titleLink)
        await waitFor(() => {
            expect(window.location.pathname).toBe(`/products/${mockProduct.id}`)
        })
    })

    it('title link click takes user to product page', () => {
        render(
            <CartProductItem
                product={mockProduct}
                handleDeleteFromCart={() => {}}
            />,
            { wrapper: BrowserRouter },
        )

        expect(
            screen.getByTestId('cart-product-item__quantity').textContent,
        ).toBe(`Qty: ${mockProduct.quantity}`)
    })

    it('trash button click calls handleDeleteFromCart', async () => {
        const user = userEvent.setup()
        const handleDeleteFromCartMock = vi.fn()
        render(
            <CartProductItem
                product={mockProduct}
                handleDeleteFromCart={handleDeleteFromCartMock}
            />,
            { wrapper: BrowserRouter },
        )
        const trashBtn = screen.getByTestId('cart-product-item__trash-btn')

        await user.click(trashBtn)

        expect(handleDeleteFromCartMock).toHaveBeenCalled()
    })
})
