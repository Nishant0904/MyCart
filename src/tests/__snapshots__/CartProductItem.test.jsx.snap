// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart Product Item test > renders correctly 1`] = `
<li
  class="sc-guDLey kdAVyx"
  data-testid="cart-product-item"
>
  <a
    data-testid="cart-product-item__image-link"
    href="/products/1"
  >
    <img
      alt="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
      src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
    />
  </a>
  <div
    class="sc-dmyCSP hIqMka"
  >
    <div>
      <a
        data-testid="cart-product-item__title-link"
        href="/products/1"
      >
        <h4>
          Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
        </h4>
      </a>
      <p>
        $109.95
      </p>
    </div>
    <div
      class="sc-hLQSwg gYnMBA"
    >
      <p
        data-testid="cart-product-item__quantity"
      >
        Qty: 
        1
      </p>
      <button
        class="sc-beySPh AktPX"
      >
        <img
          alt="trash icon"
          data-testid="cart-product-item__trash-btn"
          src="/src/assets/icons/trash-2.svg"
          title="Delete all of this item"
        />
      </button>
    </div>
  </div>
</li>
`;
